Multiple definitions of cxx_log

I'm not sure why this is not picked up in master, but I'm getting this error on a branch were I was doing some cmake changes.

The function cxx_log is declared in `physics_share.cpp` with C linkage (so no namespace qualification), to make it Fortran interoperable. However, in Homme, we also declare a cxx_log function with C linkage, in `BfbUtils.cpp`. I am not sure how these two are not already colliding in master. It looks to me as if they should.

I'm going to investigate this a bit more.

Ok, I was able to trigger the erro in master as well. But the way it triggers it raises more questions.

In the `homme_shoc_cld_p3_rrtmgp` test, we have this list of libs
```
set (NEED_LIBS cld_fraction shoc p3 scream_rrtmgp ${dynLibName} scream_control scream_share physics_share diagnostics)
```
Moving dyn lib to the front, is not enough to trigger the error, but if in addition, I remove `physics_share`, I do get the link error:
```
/home/lbertag/workdir/bin/gdb/ld: ../../../../src/physics/share/libphysics_share.a(physics_share.cpp.o): in function `cxx_log':
/home/lbertag/workdir/scream/scream-src/branch/components/eamxx/src/physics/share/physics_share.cpp:98: multiple definition of `cxx_log'; ../../../../src/dynamics/homme/tests/libtheta-l_kokkos_4_72_10.a(BfbUtils.cpp.o):/home/lbertag/workdir/scream/scream-src/branch/externals/ekat/extern/kokkos/core/src/impl/Kokkos_ViewMapping.hpp:3122: first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/CMakeFiles/homme_shoc_cld_p3_rrtmgp.dir/build.make:131: tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/homme_shoc_cld_p3_rrtmgp] Error 1
make[1]: *** [CMakeFiles/Makefile2:7164: tests/coupled/dynamics_physics/homme_shoc_cld_p3_rrtmgp/CMakeFiles/homme_shoc_cld_p3_rrtmgp.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```
Notice that the place listed as "first defined here" is completely off (there's no mention of `cxx_log` in `Kokkos_ViewMapping.hpp`, as one would expect).

What puzzles me is that removing `physics_share` from the list of libs should have no impact, since it's already brought in by p3/shoc/rrtmgp, which all link against it. I'm also puzzled by the fact that, rm-ing `physics_share` in the original list of libs does _not_ trigger the error; one needs the dyn lib to appear "early on" in the list of libs to link, more precisely before the physics libs (which are the ones bringing in physics_share). But then, if linking the dyn lib after the phys libs causes double definition (rightfully so), why doesn't linking it after also cause double definition?

@jgfouca @tcclevenger @jeff-cohere @ambrad pinging you in hope that your C++/CMake expertise could help shed some light on this nonsense...

--- a/components/eam/src/physics/cam/bfb_math.inc
+++ b/components/eam/src/physics/cam/bfb_math.inc
@@ -6,8 +6,8 @@
 ! Make sure to place the following lines at the top of any modules
 ! that use these macros:
 !
-! use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-!                              cxx_log10, cxx_exp, cxx_tanh, cxx_erf
+! use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+!                              scream_log10, scream_exp, scream_tanh, scream_erf
 
 #ifndef SCREAM_BFB_MATH_INC
 #define SCREAM_BFB_MATH_INC
@@ -30,16 +30,16 @@
 #  define bfb_tanh(val) tanh(val)
 #  define bfb_erf(val) erf(val)
 #else
-#  define bfb_pow(base, exp) cxx_pow(base, exp)
-#  define bfb_sqrt(base) cxx_sqrt(base)
-#  define bfb_cbrt(base) cxx_cbrt(base)
-#  define bfb_gamma(val) cxx_gamma(val)
-#  define bfb_log(val) cxx_log(val)
-#  define bfb_log10(val) cxx_log10(val)
-#  define bfb_exp(val) cxx_exp(val)
-#  define bfb_expm1(val) cxx_expm1(val)
-#  define bfb_tanh(val) cxx_tanh(val)
-#  define bfb_erf(val) cxx_erf(val)
+#  define bfb_pow(base, exp) scream_pow(base, exp)
+#  define bfb_sqrt(base) scream_sqrt(base)
+#  define bfb_cbrt(base) scream_cbrt(base)
+#  define bfb_gamma(val) scream_gamma(val)
+#  define bfb_log(val) scream_log(val)
+#  define bfb_log10(val) scream_log10(val)
+#  define bfb_exp(val) scream_exp(val)
+#  define bfb_expm1(val) scream_expm1(val)
+#  define bfb_tanh(val) scream_tanh(val)
+#  define bfb_erf(val) scream_erf(val)
 #endif
 
 #endif

--- a/components/eam/src/physics/cam/shoc.F90
+++ b/components/eam/src/physics/cam/shoc.F90
@@ -17,8 +17,8 @@ module shoc
 
 ! Bit-for-bit math functions.
 #ifdef SCREAM_CONFIG_IS_CMAKE
-  use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-                               cxx_log10, cxx_exp, cxx_erf
+  use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+                               scream_log10, scream_exp, scream_erf
 #endif
 
 implicit none

--- a/components/eam/src/physics/cam/trb_mtn_stress.F90
+++ b/components/eam/src/physics/cam/trb_mtn_stress.F90
@@ -5,8 +5,8 @@ module trb_mtn_stress
 
   ! Bit-for-bit math functions.
 #ifdef SCREAM_CONFIG_IS_CMAKE
-  use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-                               cxx_log10, cxx_exp, cxx_erf
+  use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+                               scream_log10, scream_exp, scream_erf
 #endif
 
   implicit none

--- a/components/eam/src/physics/cam/wv_sat_scream.F90
+++ b/components/eam/src/physics/cam/wv_sat_scream.F90
@@ -14,8 +14,8 @@ module wv_sat_scream
   use physics_utils,  only: rtype
   use micro_p3_utils, only: T_zerodegc
 #ifdef SCREAM_CONFIG_IS_CMAKE
-  use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-                                 cxx_log10, cxx_exp, cxx_tanh
+  use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+                               scream_log10, scream_exp, scream_tanh
 #endif
 
 

--- a/components/eam/src/physics/p3/eam/micro_p3.F90
+++ b/components/eam/src/physics/p3/eam/micro_p3.F90
@@ -72,8 +72,8 @@ module micro_p3
 
   ! Bit-for-bit math functions.
 #ifdef SCREAM_CONFIG_IS_CMAKE
-  use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-                                 cxx_log10, cxx_exp, cxx_expm1, cxx_tanh
+  use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+                               scream_log10, scream_exp, scream_expm1, scream_tanh
 #endif
 
   implicit none

--- a/components/eam/src/physics/p3/scream/micro_p3.F90
+++ b/components/eam/src/physics/p3/scream/micro_p3.F90
@@ -60,8 +60,8 @@ module micro_p3
 
   ! Bit-for-bit math functions.
 #ifdef SCREAM_CONFIG_IS_CMAKE
-  use physics_share_f2c, only: cxx_pow, cxx_sqrt, cxx_cbrt, cxx_gamma, cxx_log, &
-                                 cxx_log10, cxx_exp, cxx_expm1, cxx_tanh
+  use physics_share_f2c, only: scream_pow, scream_sqrt, scream_cbrt, scream_gamma, scream_log, &
+                               scream_log10, scream_exp, scream_expm1, scream_tanh
 #endif
 
   implicit none

--- a/components/eamxx/src/physics/share/physics_share.cpp
+++ b/components/eamxx/src/physics/share/physics_share.cpp
@@ -22,7 +22,7 @@ struct CudaWrap
   using Scalar = ScalarT;
   using RangePolicy = typename ekat::KokkosTypes<DeviceT>::RangePolicy;
 
-  static Scalar cxx_pow(Scalar base, Scalar exp)
+  static Scalar pow(Scalar base, Scalar exp)
   {
     Scalar result;
     RangePolicy policy(0,1);
@@ -43,106 +43,106 @@ static Scalar wrap_name(Scalar input) {                                       \
   return result;                                                              \
 }
 
-  cuda_wrap_single_arg(cxx_gamma, std::tgamma)
-  cuda_wrap_single_arg(cxx_sqrt, std::sqrt)
-  cuda_wrap_single_arg(cxx_cbrt, std::cbrt)
-  cuda_wrap_single_arg(cxx_log, std::log)
-  cuda_wrap_single_arg(cxx_log10, std::log10)
-  cuda_wrap_single_arg(cxx_exp, std::exp)
-  cuda_wrap_single_arg(cxx_expm1, std::expm1)
-  cuda_wrap_single_arg(cxx_tanh, std::tanh)
-  cuda_wrap_single_arg(cxx_erf, std::erf)
+  cuda_wrap_single_arg(gamma, std::tgamma)
+  cuda_wrap_single_arg(sqrt, std::sqrt)
+  cuda_wrap_single_arg(cbrt, std::cbrt)
+  cuda_wrap_single_arg(log, std::log)
+  cuda_wrap_single_arg(log10, std::log10)
+  cuda_wrap_single_arg(exp, std::exp)
+  cuda_wrap_single_arg(expm1, std::expm1)
+  cuda_wrap_single_arg(tanh, std::tanh)
+  cuda_wrap_single_arg(erf, std::erf)
 
 #undef cuda_wrap_single_arg
 };
 
 extern "C" {
 
-Real cxx_pow(Real base, Real exp)
+Real scream_pow(Real base, Real exp)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_pow(base, exp);
+  return CudaWrap<Real, DefaultDevice>::pow(base, exp);
 #else
   return std::pow(base, exp);
 #endif
 }
 
-Real cxx_gamma(Real input)
+Real scream_gamma(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_gamma(input);
+  return CudaWrap<Real, DefaultDevice>::gamma(input);
 #else
   return std::tgamma(input);
 #endif
 }
 
-Real cxx_cbrt(Real input)
+Real scream_cbrt(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_cbrt(input);
+  return CudaWrap<Real, DefaultDevice>::cbrt(input);
 #else
   return std::cbrt(input);
 #endif
 }
 
-Real cxx_sqrt(Real input)
+Real scream_sqrt(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_sqrt(input);
+  return CudaWrap<Real, DefaultDevice>::sqrt(input);
 #else
   return std::sqrt(input);
 #endif
 }
 
-Real cxx_log(Real input)
+Real scream_log(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_log(input);
+  return CudaWrap<Real, DefaultDevice>::log(input);
 #else
   return std::log(input);
 #endif
 }
 
-Real cxx_log10(Real input)
+Real scream_log10(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_log10(input);
+  return CudaWrap<Real, DefaultDevice>::log10(input);
 #else
   return std::log10(input);
 #endif
 }
 
-Real cxx_exp(Real input)
+Real scream_exp(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_exp(input);
+  return CudaWrap<Real, DefaultDevice>::exp(input);
 #else
   return std::exp(input);
 #endif
 }
 
-Real cxx_expm1(Real input)
+Real scream_expm1(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_expm1(input);
+  return CudaWrap<Real, DefaultDevice>::expm1(input);
 #else
   return std::expm1(input);
 #endif
 }
   
-Real cxx_tanh(Real input)
+Real scream_tanh(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_tanh(input);
+  return CudaWrap<Real, DefaultDevice>::tanh(input);
 #else
   return std::tanh(input);
 #endif
 }
 
-Real cxx_erf(Real input)
+Real scream_erf(Real input)
 {
 #ifdef EAMXX_ENABLE_GPU
-  return CudaWrap<Real, DefaultDevice>::cxx_erf(input);
+  return CudaWrap<Real, DefaultDevice>::erf(input);
 #else
   return std::erf(input);
 #endif

--- a/components/eamxx/src/physics/share/physics_share.hpp
+++ b/components/eamxx/src/physics/share/physics_share.hpp
@@ -7,16 +7,16 @@ namespace scream {
 
 extern "C" {
 
-Real cxx_pow(Real base, Real exp);
-Real cxx_sqrt(Real base);
-Real cxx_cbrt(Real base);
-Real cxx_gamma(Real input);
-Real cxx_log(Real input);
-Real cxx_log10(Real input);
-Real cxx_exp(Real input);
-Real cxx_expm1(Real input);
-Real cxx_tanh(Real input);
-Real cxx_erf(Real input);
+Real scream_pow(Real base, Real exp);
+Real scream_sqrt(Real base);
+Real scream_cbrt(Real base);
+Real scream_gamma(Real input);
+Real scream_log(Real input);
+Real scream_log10(Real input);
+Real scream_exp(Real input);
+Real scream_expm1(Real input);
+Real scream_tanh(Real input);
+Real scream_erf(Real input);
 
 }
 

--- a/components/eamxx/src/physics/share/physics_share_f2c.F90
+++ b/components/eamxx/src/physics/share/physics_share_f2c.F90
@@ -21,7 +21,7 @@ interface
   ! the C++ versions in order to stay BFB.
   !
 
-  function cxx_pow(base, exp) bind(C)
+  function scream_pow(base, exp) bind(C)
     use iso_c_binding
 
     !arguments:
@@ -29,98 +29,98 @@ interface
     real(kind=c_real), value, intent(in)  :: exp
 
     ! return
-    real(kind=c_real)               :: cxx_pow
-  end function cxx_pow
+    real(kind=c_real)               :: scream_pow
+  end function scream_pow
 
-  function cxx_sqrt(base) bind(C)
+  function scream_sqrt(base) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in)  :: base
 
     ! return
-    real(kind=c_real)               :: cxx_sqrt
-  end function cxx_sqrt
+    real(kind=c_real)               :: scream_sqrt
+  end function scream_sqrt
 
-  function cxx_cbrt(base) bind(C)
+  function scream_cbrt(base) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in)  :: base
 
     ! return
-    real(kind=c_real)               :: cxx_cbrt
-  end function cxx_cbrt
+    real(kind=c_real)               :: scream_cbrt
+  end function scream_cbrt
 
-  function cxx_gamma(input) bind(C)
+  function scream_gamma(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_gamma
-  end function cxx_gamma
+    real(kind=c_real)            :: scream_gamma
+  end function scream_gamma
 
-  function cxx_log(input) bind(C)
+  function scream_log(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_log
-  end function cxx_log
+    real(kind=c_real)            :: scream_log
+  end function scream_log
 
-  function cxx_log10(input) bind(C)
+  function scream_log10(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_log10
-  end function cxx_log10
+    real(kind=c_real)            :: scream_log10
+  end function scream_log10
 
-  function cxx_exp(input) bind(C)
+  function scream_exp(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_exp
-  end function cxx_exp
+    real(kind=c_real)            :: scream_exp
+  end function scream_exp
 
-  function cxx_expm1(input) bind(C)
+  function scream_expm1(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_expm1
-  end function cxx_expm1
+    real(kind=c_real)            :: scream_expm1
+  end function scream_expm1
   
-  function cxx_tanh(input) bind(C)
+  function scream_tanh(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in) :: input
 
     ! return
-    real(kind=c_real)            :: cxx_tanh
-  end function cxx_tanh
+    real(kind=c_real)            :: scream_tanh
+  end function scream_tanh
 
-  function cxx_erf(input) bind(C)
+  function scream_erf(input) bind(C)
     use iso_c_binding
 
     !arguments:
     real(kind=c_real), value, intent(in)  :: input
 
     ! return
-    real(kind=c_real) :: cxx_erf
-  end function cxx_erf
+    real(kind=c_real) :: scream_erf
+  end function scream_erf
 
 end interface
 

